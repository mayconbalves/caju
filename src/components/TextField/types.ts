export type TextFieldProps = {
  id?: string
  label?: string
  error?: string
  maxLength?: number
  name: string
  onChange: (e: React.ChangeEvent<HTMLInputElement>) => void
  placeholder?: string
  type?: string
  value: string
}

export type InputStylesProps = {
  error?: string
}
