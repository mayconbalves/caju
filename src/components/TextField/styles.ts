import styled from 'styled-components'
import { InputStylesProps } from './types'

export const Wrapper = styled.div`
  display: flex;
  flex-direction: column;
  gap: 5px;
`

export const Input = styled.input<InputStylesProps>`
  padding: 0 8px;
  vertical-align: middle;
  border-radius: 2px;
  min-height: 36px;
  background-color: #ffffff;
  border: 1px solid rgba(36, 28, 21, 0.3);
  transition: all 0.2s ease-in-out 0s;
  font-size: 16px;
  line-height: 18px;
  font-weight: normal;
  border-radius: 8px;

  ${({ error }) =>
    error
      ? `
    outline: none;
    border: 2px solid rgba(232, 5, 55, 1)
  `
      : `
    &:focus {
      outline: none;
      border: 1px solid #007c89;
      box-shadow: inset 0 0 0 1px #007c89;
    }
  `}
`

export const Hint = styled.span`
  color: rgba(232, 5, 55, 1);
  font-size: small;
`
