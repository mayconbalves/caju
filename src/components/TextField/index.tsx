import { Hint, Input, Wrapper } from './styles'
import { TextFieldProps } from './types'

const TextField = ({
  error,
  id,
  maxLength,
  name,
  onChange,
  placeholder,
  label,
  type,
  value
}: TextFieldProps) => {
  return (
    <Wrapper>
      <label htmlFor={id}>{label}</label>
      <Input
        aria-label={label}
        aria-describedby={error}
        aria-invalid={!!error}
        error={error}
        maxLength={maxLength}
        name={name}
        onChange={onChange}
        placeholder={placeholder}
        type={type}
        value={value}
      />
      {error && <Hint id={`${id}-error`}>{error}</Hint>}
    </Wrapper>
  )
}

export default TextField
