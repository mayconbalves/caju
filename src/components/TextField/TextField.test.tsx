import { fireEvent, render, screen } from '@testing-library/react'
import TextField from '.'

describe('TextField Component', () => {
  it('should render with label and placeholder', () => {
    render(
      <TextField
        id="name"
        name="name"
        label="Name"
        placeholder="Enter your name"
        value=""
        onChange={() => {}}
      />
    )

    expect(screen.getByLabelText('Name')).toBeInTheDocument()
    expect(screen.getByPlaceholderText('Enter your name')).toBeInTheDocument()
  })

  it('should display an error message when error is present', () => {
    render(
      <TextField
        id="name"
        name="name"
        label="Name"
        placeholder="Enter your name"
        value=""
        error="Name is required"
        onChange={() => {}}
      />
    )

    expect(screen.getByText('Name is required')).toBeInTheDocument()
    expect(screen.getByRole('textbox', { name: 'Name' })).toHaveAttribute('aria-invalid', 'true')
  })

  it('should call onChange when typing in the input', () => {
    const handleChange = jest.fn()
    const { container } = render(
      <TextField
        id="name"
        name="name"
        label="Name"
        placeholder="Enter your name"
        value=""
        onChange={handleChange}
      />
    )

    const input = screen.getByPlaceholderText('Enter your name')
    fireEvent.change(input, { target: { value: 'John Doe' } })
    expect(handleChange).toHaveBeenCalledTimes(1)
    expect(container).toMatchSnapshot()
  })
})
