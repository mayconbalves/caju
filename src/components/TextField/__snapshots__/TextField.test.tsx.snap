// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextField Component should call onChange when typing in the input 1`] = `
<div>
  <div
    class="sc-aYaIB dmBzed"
  >
    <label
      for="name"
    >
      Name
    </label>
    <input
      aria-invalid="false"
      aria-label="Name"
      class="sc-gEvDqW hdDevj"
      name="name"
      placeholder="Enter your name"
      value=""
    />
  </div>
</div>
`;
